src/pybind/mgr/cephadm/schedule.py

```diff
@@ -12,18 +12,76 @@
     Protocol,
     Tuple,
     TypeVar,
+    TYPE_CHECKING,
 )
 
 import orchestrator
-from ceph.deployment.service_spec import ServiceSpec
+from ceph.deployment.service_spec import ServiceSpec, HostPlacementSpec
 from orchestrator._interface import DaemonDescription
 from orchestrator import OrchestratorValidationError
 from .utils import RESCHEDULE_FROM_OFFLINE_HOSTS_TYPES
+if TYPE_CHECKING:
+    from cephadm.module import CephadmOrchestrator
 
 logger = logging.getLogger(__name__)
 T = TypeVar('T')
 
 
+def get_placement_hosts(
+    spec: ServiceSpec,
+    hosts: List[orchestrator.HostSpec] = [],
+    draining_hosts: List[orchestrator.HostSpec] = [],
+    mgr: Optional['CephadmOrchestrator'] = None
+) -> List[HostPlacementSpec]:
+    """
+    Get the list of candidate host placement specs based on placement specifications.
+    Args:
+        spec: The service specification
+        hosts: List of available hosts (optional, will be generated from mgr if not provided)
+        draining_hosts: List of hosts that are draining (optional, will be generated from mgr if not provided)
+        mgr: CephadmOrchestrator instance to generate hosts from if provided
+    Returns:
+        List[HostPlacementSpec]: List of host placement specs that match the placement criteria
+    """
+    if not hosts and mgr:
+        hosts = mgr.cache.get_schedulable_hosts()
+    if not draining_hosts and mgr:
+        draining_hosts = mgr.cache.get_draining_hosts()
```

Comment:
we've been able to avoid having to ever import the `CephadmOrchestrator` / `mgr` here up until now. I'd prefer if we just allow this function to take a `hosts`, `draining_hosts`, `schedulable_hosts` params that can then be passed in from anywhere that would call this. Everywhere we would should either already have access to the mgr and therefore can provide them with the exception of the one call within schedule.py, where the `HostAssignment` class it's being called from should have already been provided with those lists